celery/canvas.py

```diff
@@ -1221,20 +1222,42 @@ def freeze(self, _id=None, group_id=None, chord=None,
             opts['group_index'] = group_index
         root_id = opts.setdefault('root_id', root_id)
         parent_id = opts.setdefault('parent_id', parent_id)
-        new_tasks = []
-        # Need to unroll subgroups early so that chord gets the
-        # right result instance for chord_unlock etc.
-        results = list(self._freeze_unroll(
-            new_tasks, group_id, chord, root_id, parent_id,
-        ))
-        if isinstance(self.tasks, MutableSequence):
-            self.tasks[:] = new_tasks
+        if isinstance(self.tasks, _regen):
+            # We are draining from a geneator here.
+            tasks1, tasks2 = itertools.tee(self._unroll_tasks(self.tasks))
+            results = regen(self._freeze_tasks(tasks1, group_id, chord, root_id, parent_id))
+            self.tasks = regen(x[0] for x in zip(tasks2, results))
```

Comment:
Hi @mjhoffman65 , I'm a new project member. 
I am trying to understand what you did here, I would appreciate if you could clarify it for me. 
I don't get why you zipped `tasks2` with `results` only to get `x[0]`, resulting in a generator with only `tasks2`'s tasks. Why did you write this and not `self.tasks = regen(tasks2)`?
TIA :)